modules/core/src/core/animation-loop.js

```diff
@@ -101,49 +101,54 @@ export default class AnimationLoop {
 
   // Starts a render loop if not already running
   // @param {Object} context - contains frame specific info (E.g. tick, width, height, etc)
-  start(opts = {}) {
+  async start(opts = {}) {
     if (this._running) {
-      return this;
+      return;
     }
     this._running = true;
+
     // console.debug(`Starting ${this.constructor.name}`);
     // Wait for start promise before rendering frame
-    this._startPromise = getPageLoadPromise()
-      .then(() => {
-        if (!this._running || this._initialized) {
-          return null;
-        }
-
-        // Create the WebGL context
-        this._createWebGLContext(opts);
-        this._createFramebuffer();
-        this._startEventHandling();
-
-        // Initialize the callback data
-        this._initializeCallbackData();
-        this._updateCallbackData();
-
-        // Default viewport setup, in case onInitialize wants to render
-        this._resizeCanvasDrawingBuffer();
-        this._resizeViewport();
-
-        // Note: onIntialize can return a promise (in case it needs to load resources)
-        const initializationPromise = this.onInitialize(this.animationProps);
-        this._initialized = true;
-        return initializationPromise;
-      })
-      .then(appContext => {
-        if (this._running) {
-          this._addCallbackData(appContext || {});
-          if (appContext !== false) {
-            this._startLoop();
-          }
-        }
-      });
+    await getPageLoadPromise();
+
+    if (!this._running || this._initialized) {
+      return;
```

Comment:
I think we're introducing a bug here. If I'm reading this correctly, it looks like once `this._initialized === true`, the user will be unable to call `stop()` and `start()` again. In the previous code the `return` would just drop out of the callback containing the initialization code but still run the next callback that actually starts the render loop. I think what we want to do is stick all these initialization calls in an `if (this._running && !this._initialized)` block.

Reply:
When I first saw this, I wondered why the `start(); stop(); start()` test didn't fail. Looking at the test more closely, I realized that we're only testing that the initialization functions were only called once. We don't check that the render loop actually starts up when calling `start()` after a `stop()`. Maybe we should add that check to the test?